power_grid_model_c/power_grid_model/include/power_grid_model/main_model.hpp

```diff
@@ -281,6 +296,14 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
         }
     }
 
+    // restore the initial values of all components
+    void restore_components() {
+        components_.restore_values();
+        is_topology_up_to_date_ = false;
+        is_sym_parameter_up_to_date_ = false;
+        is_asym_parameter_up_to_date_ = false;
```

Comment:
Needs help: this kills performance.

Possible solutions:

* [ ] Also cache and restore the topology
* [ ] Remember whether the `UpdateChange` mutated topology
  * This may cause repeatedly recalculate the same topology
* [ ] Completely different approach: Instead of caching components, cache the inverse patches/mutations and define operation `patch_y o patch_x_inverse`
  * E.g.
    * Input: component A `{.p = a, .q = b}`
    * Scenario 1: update component A `{.p = c}`
      * The inverse patch is `{.p = a}`
    * Scenario 2: update component A `{.q = d}`
      * The inverse patch is `{.q = b}`
    * Parallel: OK
    * Sequential (1 -> 2): `{.p = a, .q = b}`
      * Scenario 1:
        * apply patch `(1) o (inverse noOp) = {.p = c}` -> `{.p = c}`
        * check topology changed etc.
        * run calculation
      * Scenario 2:
        * apply combined patch `(2) o (inverse 1) = {.p = a, .q = d}` -> `{.p = a, .q = d}`
        * check topology changed etc.
        * run calculation
  * This way, topology is diffed against the latest run, not against the main topology. This could be useful if e.g. an entire batch of scenarios is run that has the same topology within the batch, but that topology is different from the input state

Reply:
@TonyXiang8787 I updated the code with your comments. This question, however, still stands and requires your input.

Reply:
cfr. call: go with option 2.

The user should be smart enough to think about which types of calculations to do with a single permanent update (e.g.: a set of simulations with a the same topology that is different from that of the `0`-case).

In the future, we may want to implement a "smart dispatch" which batches for these kind of expensive changes like updating topology, but out of scope of this PR